packages/flutter/lib/src/widgets/scroll_physics.dart

```diff
@@ -328,6 +376,49 @@ class ScrollPhysics {
   }
 }
 
+/// Scroll physics that attempt to keep the scroll position in range when the
+/// contents change dimensions suddenly.
+///
+/// If the scroll position is already out of range, this attempts to maintain
+/// the amount of overscroll or underscroll already present.
+///
+/// If the scroll activity is animating the scroll position, sudden changes to
+/// the scroll dimensions are allowed to happen (so as to prevent animations
+/// from jumping back and forth between in-range and out-of-range values).
+///
+/// These physics should be combined with other scroll physics, e.g.
+/// [BouncingScrollPhysics] or [ClampingScrollPhysics], to obtain a complete
+/// description of typical scroll physics. See [applyTo].
+class RangeMaintainingScrollPhysics extends ScrollPhysics {
+  /// Creates scroll physics that maintain the scroll position in range.
+  const RangeMaintainingScrollPhysics({ ScrollPhysics parent }) : super(parent: parent);
+
+  @override
+  RangeMaintainingScrollPhysics applyTo(ScrollPhysics ancestor) {
+    return RangeMaintainingScrollPhysics(parent: buildParent(ancestor));
+  }
+
+  @override
+  double adjustPositionForNewDimensions({
+    @required ScrollMetrics oldPosition,
+    @required ScrollMetrics newPosition,
+    @required bool isScrolling,
+    @required double velocity,
+  }) {
+    if (velocity != 0.0 || ((oldPosition.minScrollExtent == newPosition.minScrollExtent) && (oldPosition.maxScrollExtent == newPosition.maxScrollExtent)))
+      return super.adjustPositionForNewDimensions(oldPosition: oldPosition, newPosition: newPosition, isScrolling: isScrolling, velocity: velocity);
+    if (oldPosition.pixels < oldPosition.minScrollExtent) {
+      final double oldDelta = oldPosition.minScrollExtent - oldPosition.pixels;
+      return newPosition.minScrollExtent - oldDelta;
```

Comment:
What is the logic here?  If the old position is out of bounds that means that it wasn't previously corrected right?  Then why do we take it into consideration?  There is some weird feedback happening where the last decision made by this function is affecting its next decision, no?

Reply:
The most common case for why you'd be out of bounds here is that the user was previously overscrolling (e.g. on iOS). It also might happen in cases where the initial scroll position is out of bounds, though I didn't run into that case in testing so that's just a guess.